storage/grpc_client.go

```diff
@@ -1652,38 +1664,43 @@ func (w *gRPCWriter) uploadBuffer(recvd int, start int64, doneReading bool) (*st
 				// on the *last* message of the stream (instead of the first).
 				req.ObjectChecksums = toProtoChecksums(w.sendCRC32C, w.attrs)
 			}
-
 		}
 
 		err = w.stream.Send(req)
 		if err == io.EOF {
 			// err was io.EOF. The client-side of a stream only gets an EOF on Send
 			// when the backend closes the stream and wants to return an error
 			// status. Closing the stream receives the status as an error.
-			_, err = w.stream.CloseAndRecv()
+			err = w.stream.CloseSend()
+
+			// Drop the stream reference as a new one will need to be created
+			w.stream = nil
```

Comment:
Are we 100% sure we can drop the stream reference here? If `w.stream.Send` gets an io.EOF, it means the backend wants to send an error. The error status won't come from CloseSend(), I think the client side still needs to `Recv` to get the actual Status (will send a link offline).  I think we need to call Recv or CloseAndRecv if the bidi stream still has it, before Closing.

Reply:
Thanks Noah, you are 100% correct. In this case, it would be a Recv() (the bidi stream does not have CloseAndRecv). Updated the code.

As for testing it out in practice, I agree. I think the best is to do this via the conformance tests, pending testbench implementation.